terrarium/bi_ci/bi_ci/execute_mypy_multi.py

```diff
@@ -13,52 +16,155 @@
+        cwd=str(request.pkg_dir),
+    )
+    stdout, stderr = await proc.communicate()
+    await cache_queue.put(mypy_cache_dir)
+    return MypyResult(
+        pkg_dir=request.pkg_dir,
+        command=command,
+        status=MypyStatus.SUCCESS if proc.returncode == 0 else MypyStatus.FAIL,
+        stdout=stdout,
+        stderr=stderr,
+    )
+
+
+def main(root: Path, targets_file: Path = None, processes: int = 1) -> None:  # type: ignore  # 2024-01-30 # TODO: Incompatible default for argument "targets_file" (default has type "None", argument has type "Path")  [assignment]
+    """
+    Run mypy on the datalens backend code.
+    We have to run mypy separately on each project
+    because mypy reads its settings from pyproject.toml from its current working directory.
+
+    This script DON'T check mypy in tests.
+    """
     # clize can't recognize type annotation "Optional"
-    paths: Iterable[str]
+    package_rel_paths: Iterable[str]
     if targets_file is not None:
-        paths = json.load(open(targets_file))
+        package_rel_paths = json.load(open(targets_file))
     else:
-        paths = get_targets(root)
-    failed_list: list[str] = []
-    mypy_cache_dir = Path("/tmp/mypy_cache")
-    mypy_cache_dir.mkdir(exist_ok=True)
-    for path in paths:
-        pkg_dir = root / path
-        run_args = ["mypy", f"--cache-dir={mypy_cache_dir}"]
+        package_rel_paths = get_python_packages(root)
+
+    base_mypy_cache_dir = Path("/tmp/mypy_cache")
+    base_mypy_cache_dir.mkdir(exist_ok=True)
+
+    mypy_requests: list[MypyRequest] = []
+    for rel_path in package_rel_paths:
+        print()  # just separator
+        pkg_dir = root / rel_path
         targets = get_mypy_targets(pkg_dir)
-        print(f"Cmd: {run_args}; cwd={pkg_dir}")
-        if len(targets) > 0:
-            run_args.extend(targets)
-            run_exit_code = subprocess.run(" ".join(run_args), shell=True, cwd=str(pkg_dir)).returncode
-            if run_exit_code != 0:
-                failed_list.append(path)
-        else:
-            print(f"mypy config not found in {pkg_dir}/pyproject.toml, skipped")
-
-    if failed_list:
-        print("Mypy failed for targets:")
-        print("\n".join(sorted(failed_list)))
-
-    sys.exit(0 if len(failed_list) == 0 else 1)
+        if len(targets) == 0:
+            print(f"SKIP: there weren't any valid targets for {pkg_dir}")
+            continue
+        print(f"PLAN: process {pkg_dir} with {targets}")
+        mypy_requests.append(
+            MypyRequest(
+                pkg_dir=pkg_dir,
+                targets=targets,
+            )
+        )
+
+    cache_queue = CacheQueue(
+        base_mypy_cache_dir=base_mypy_cache_dir,
+        count=processes,
+    )
+    loop = asyncio.get_event_loop()
+    results: Iterable[MypyResult] = loop.run_until_complete(
+        asyncio.gather(*[run_mypy(request=request, cache_queue=cache_queue) for request in mypy_requests])
+    )
```

Comment:
asyncio.run(...) or add loop.close afterwards and let's move tasks preparation on different line for cleaner code